fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java

```diff
@@ -332,40 +336,132 @@ public <M extends Message> CompletableFuture<Void> update(@Nullable FDBIndexable
+
+    /**
+     * remove a deleted document from its partition metadata.
+     *
+     * @param oldRecord record to be deleted
+     * @param groupingKey grouping key
+     * @return partition id (or null if partitioning isn't enabled)
+     */
+    private <M extends Message> Integer removeDeletedDocumentFromPartition(@Nullable FDBIndexableRecord<M> oldRecord, @Nonnull Tuple groupingKey) {
+        if (oldRecord == null || !partitioner.isPartitioningEnabled()) {
+            return null;
+        }
+
+        Long oldRecordTimestamp = getPartitioningTimestampValue(partitioner.getPartitionTimestampFieldName(), oldRecord);
+
+        // find the document's partition
+        LucenePartitionInfoProto.LucenePartitionInfo partition = partitioner.locateInPartition(groupingKey, oldRecordTimestamp);
+        if (partition == null) {
+            throw new RecordCoreException("Issue deleting old index keys (partition not found)", "oldRecord", oldRecord);
+        }
+        partitioner.removeFromAndSavePartitionMetadata(groupingKey, partition);
+        return partition.getId();
+    }
+
+    private <M extends Message> Long getPartitioningTimestampValue(@Nonnull String fieldName, @Nullable FDBIndexableRecord<M> rec) {
+        if (rec == null) {
+            return null;
+        }
+
+        int dotIndex = fieldName.indexOf('.');
+        String containingTypeName = null;
+        String actualFieldName = fieldName;
+        if (dotIndex >= 0) {
+            // nested
+            containingTypeName = fieldName.substring(0, dotIndex);
+            actualFieldName = fieldName.substring(dotIndex + 1);
+        }
+        if (containingTypeName != null) {
+            for (Map.Entry<Descriptors.FieldDescriptor, Object> field : rec.getRecord().getAllFields().entrySet()) {
+                if (containingTypeName.equalsIgnoreCase(field.getKey().getName()) && field.getValue() instanceof Message) {
+                    return getPartitioningTimestampValue(actualFieldName, ((Message) field.getValue()).getAllFields());
+                }
             }
-        } catch (IOException e) {
-            throw new RecordCoreException("Issue updating new index keys", e)
-                    .addLogInfo("newRecord", newRecord);
+        } else {
+            return getPartitioningTimestampValue(actualFieldName, rec.getRecord().getAllFields());
         }
+        return null;
+    }
 
-        return AsyncUtil.DONE;
+    private Long getPartitioningTimestampValue(@Nonnull String fieldName, Map<Descriptors.FieldDescriptor, Object> fields) {
+        for (Map.Entry<Descriptors.FieldDescriptor, Object> field : fields.entrySet()) {
+            if (fieldName.equalsIgnoreCase(field.getKey().getName())) {
+                if (field.getValue() instanceof Long) {
+                    return (Long)field.getValue();
+                }
+                if (field.getValue() instanceof Double) {
+                    return ((Double)field.getValue()).longValue();
+                }
```

Comment:
I added this because I was using the record definitions in the tests, where `ComplexDocument.time` is defined as `double`. I can make this only accept `long`, but I will have to tweak the test record definition (or create a new one, which may be simpler)

Reply:
It should be fine to add another field to that message, or you could use `score`. Creating a new record type is also fine.